pydantic/warnings.py

```diff
@@ -0,0 +1,45 @@
+"""Pydantic-specific warnings."""
+from __future__ import annotations as _annotations
+
+__all__ = 'PydanticDeprecatedSince20', 'PydanticDeprecationWarning'
+
+
+class PydanticDeprecationWarning(DeprecationWarning):
+    """A Pydantic specific deprecation warning.
+
+    This warning is raised when using deprecated functionality in Pydantic. It provides information on when the
+    deprecation was introduced and the expected version in which the corresponding functionality will be removed.
+
+    Attributes:
+        message: Description of the warning.
+        since: Pydantic version in what the deprecation was introduced.
+        expected_removal: Pydantic version in what the corresponding functionality expected to be removed.
+    """
+
+    message: str
+    since: tuple[int, int]
+    expected_removal: tuple[int, int]
+
+    def __init__(
+        self, message: str, *args: object, since: tuple[int, int], expected_removal: tuple[int, int] | None = None
+    ) -> None:
+        super().__init__(message, *args)
+        self.message = message.rstrip('.')
+        self.since = since
+        self.expected_removal = expected_removal if expected_removal is not None else (since[0] + 1, 0)
+
+    def __str__(self) -> str:
+        message = (
+            f'{self.message}. Deprecated in Pydantic V{self.since[0]}.{self.since[1]}'
+            f' to be removed in V{self.expected_removal[0]}.{self.expected_removal[1]}.'
+        )
+        if self.since == (2, 0):
+            message += ' See Pydantic V2 Migration Guide at https://errors.pydantic.dev/v2.0/migration/'
+        return message
```

Comment:
The integration of this warning with the IDEs is also not ideal, since the `__str__` can't be used, and I'd expect us to have that in there. This includes both VSCode, and PyCharm.

I also think that each warning should link to the specific part of the documentation instead of a generic link to the guide (I don't think we have everything covered there), and it's likely that we are going to have multiple links on the same warning.

I suggest a simple implementation instead:
```suggestion
class PydanticDeprecationWarning(DeprecationWarning):
    ...
```

Reply:
> The integration of this warning with the IDEs is also not ideal, since the __str__ can't be used, and I'd expect us to have that in there. This includes both VSCode, and PyCharm.

This is true for any custom `DeprecationWarning`. The current `PydanticDeprecationWarning` implementation is compatible with the `DeprecationWarning` in terms of the first argument being a deprecation reason provided explicitly in the code. It is true that everything else is being added in runtime.

Reply:
> I suggest a simple implementation instead

Basically, you're suggesting throwing away the entire PR. Any particular reason for that?

Reply:
> Basically, you're suggesting throwing away the entire PR. Any particular reason for that?

Well, if you don't want the IDE support of "showing the entire message", then no need to change. 🤷‍♂️

Reply:
I think it's fine for the IDE to not show the entire message.